src/audio/smart_amp/smart_amp_generic.c

```diff
@@ -9,194 +9,292 @@
 #include <sof/audio/format.h>
 #include <sof/audio/smart_amp/smart_amp.h>
 
-#ifdef CONFIG_GENERIC
+static void remap_s32_to_s32(struct smart_amp_mod_stream *src_mod, uint32_t frames,
+			     const struct audio_stream __sparse_cache *src,
+			     const int8_t *chan_map)
+{
+	int fr, ch;
+	int src_frag = 0;
+	int src_ch = audio_stream_get_channels(src);
+	int src_idx;
+	int32_t *src_ptr;
+	int32_t *mod_ptr = (int32_t *)src_mod->buf.data;
+
+	for (fr = 0 ; fr < frames ; fr++) {
+		for (ch = 0 ; ch < src_mod->channels; ch++) {
+			if (chan_map[ch] == -1) {
+				*mod_ptr = 0;
+			} else {
+				src_idx = src_frag + chan_map[ch];
+				src_ptr = audio_stream_read_frag_s32(src, src_idx);
+				*mod_ptr = *src_ptr;
+			}
+			mod_ptr++;
+		}
+		src_frag += src_ch;
+	}
+}
 
-static int32_t smart_amp_ff_generic(int32_t x)
+static void remap_s24_to_s24(struct smart_amp_mod_stream *src_mod, uint32_t frames,
+			     const struct audio_stream __sparse_cache *src,
+			     const int8_t *chan_map)
 {
-	/* Add speaker protection feed forward process here */
-	return x;
+	remap_s32_to_s32(src_mod, frames, src, chan_map);
 }
 
-static void smart_amp_fb_generic(int32_t x)
+static void remap_s24_to_s32(struct smart_amp_mod_stream *src_mod, uint32_t frames,
+			     const struct audio_stream __sparse_cache *src,
+			     const int8_t *chan_map)
 {
-	/* Add speaker protection feedback process here */
+	int fr, ch;
+	int src_frag = 0;
+	int src_ch = audio_stream_get_channels(src);
+	int src_idx;
+	int32_t *src_ptr;
+	int32_t *mod_ptr = (int32_t *)src_mod->buf.data;
+
+	for (fr = 0 ; fr < frames ; fr++) {
+		for (ch = 0 ; ch < src_mod->channels; ch++) {
+			if (chan_map[ch] == -1) {
+				*mod_ptr = 0;
+			} else {
+				src_idx = src_frag + chan_map[ch];
+				src_ptr = audio_stream_read_frag_s32(src, src_idx);
+				*mod_ptr = *src_ptr << 8;
```

Comment:
here too, use audio_stream_read_frag_s32(), there is an example in aria_algo_get_data() in aria_generic.c